### ISSUE TITLE
Values typed as `Function` should not be callable

### ISSUE THREAD

**Issue by lazytype** [microsoft/TypeScript#48284]
# Bug Report
### 🔎 Search Terms
builtin function interface callable constructor narrow unsound

### 🕗 Version & Regression Information
This goes back at least as far as 3.3.3.
- This is the behavior in every version I tried, and I reviewed the FAQ for entries about anything mentioning functions.

### ⏯ Playground Link
[Playground link with relevant code](https://www.typescriptlang.org/play?target=7&ts=4.6.2#code/C4TwDgpgBA8gdgGxAYQIYIagRg6BeKACgDpTUAnAcwGcAuKVOEASijwD4GmBuAWAChQkWIhQB7ONWDkArgGNgY8mgzZcbKHAgB3IqWIUa9RizacTAodACCTAGIy4CgJYSN8JCsw5oAHxGeElKyCkpeahACAgBmji5ucugIhLFO9LYgDk7ArnCsAN4CUFAA9CVQAALA1AC0EAAekAp15ORKUACyqADWENRQ1BCSkfzFqXKEzHyjpeVVtQ1NwC1t5J09fQNDg0WaOlDjk9O7ZZXVdY0QzRCt7V29-YPDu3JBwAz0HihJERrj08VTvMLksVncNo9tiNiq9JO8sJ9RMg3iFFMofj4-nFjjNTgB1AAWICgzn6wAJpM0YnejCgN1WAH4Xm8oHJ6Fl4nAsU4ccVUEcTnNzosrst6eCHltnjMtLp+VMojL9lgBbihQtLtdbmt7psnjsZiqFfxBVBCcTKeTKXBqVw6dqmTMJsbAeUAEIyd5W-qkx3FWWs1UAXyAA)

### 💻 Code
```ts
type OnlyCallable = (...args: any) => any;
type OnlyConstructorCallable = new (...args: any) => any
type AnyFunction = OnlyCallable | OnlyConstructorCallable

function call(func: AnyFunction) {
  // @ts-expect-error Makes sense
  func();
  // @ts-expect-error Makes sense
  new func();

  // @ts-expect-error Makes sense
  const a: OnlyCallable = func;
  // @ts-expect-error Makes sense
  const b: OnlyConstructorCallable = func;

  // Why is this not an error?
  const c: Function = func;

  a();
  // @ts-expect-error Makes sense
  new a();

  new b();
  // @ts-expect-error Makes sense
  b();

  // Why is this not an error?
  c();
  // But this is?
  new c();
}
```

### 🙁 Actual behavior
`Function` seems to be an unsafer version of `(...args: any) => any`, since it allows assigning `AnyFunction` to it.

### 🙂 Expected behavior
`Function` should behavior either like `(...args: any) => any` or `AnyFunction`, or even `OnlyCallable & OnlyConstructorCallable`.

What's worse is that
```ts
function foo(value: unknown) {
  if (typeof value === 'function') {
    // `value` is narrowed to `Function`
    value();
  }

  if (value instanceof Function) {
    // `value` is narrowed to `Function` 
    value();
  }
}

// @ts-expect-error Makes sense
Promise()

// No type error but this will throw at runtime
foo(Promise)
```

**Comment by RyanCavanaugh** [comment]
This is the intended behavior; the term we use is "untyped function call". It'd be possible to add a flag to turn this off but this behavior dates back to TypeScript 0.8 and I haven't seen much feedback about it yet.

**Comment by lazytype** [comment]
The heuristic of allowing untyped functions to be callable makes sense to me since the vast majority of "functions" are ones that can be called without "new". What I mostly take issue with is allowing an `OnlyConstructorCallable` and `AnyCallable` as defined above to be assigned to a `Function`. The class of errors this would allow feels similar to the class of errors that would be allowed by being able to assign a `ReadonlyArray` to an `Array`.

I think disallowing such assignments would affect very few users and also make `Function` be less of a special, magical type. The simplest way to achieve that would be to just define `Function` as `(...args: any) => any`, but I feel like that must have been proposed before. Are there good examples where that would be undesirable?

**Comment by RyanCavanaugh** [comment]
`Function` being a valid assignment target of anything callable is the non-negotiable part -- `Function` itself has the same relation to regular functions as `string` does with `String`, `number` with `Number`, etc

**Comment by fatcerberus** [comment]
`new () => T` without a standard call signature is not technically callable though - e.g. `Promise();` without `new` will throw.

**Comment by RyanCavanaugh** [comment]
Constructors still have `bind` / `call` / `apply`, which is what `Function` is there to provide

**Comment by fatcerberus** [comment]
IIRC it’s always a runtime error to invoke any of those on a class constructor.  You need to use `Reflect.construct()` to invoke them.

**Comment by RyanCavanaugh** [comment]
Yeah, it's confusing how to reason about things which are definitely present but definitely not callable (the non-user-invokable HTML element constructors in the DOM also have this problem). Technically you can call `ctor.bind.bind(...`, so it's there...

**Comment by lazytype** [comment]
I think I've managed to convince myself that the current state is ok. Since this is allowed:
```ts
function foo(numberArray: number[]) {
  // IIRC this works because methods are bivariant (will #48240 allow more control over this?)
  const numberOrStringArray: Array<number | string> = numberArray;
  numberOrStringArray.push('hello');
}
```

I guess the asymmetry between calls with and without new still kind of bugs me but I don't see much practical benefit of allowing a `Function` to be called with `new`.

Feel free to close or leave open for discussion.

**Comment by Ginden** [comment]
> IIRC it’s always a runtime error to invoke any of those on a class constructor. 

`.bind` can be safely used with class constructors. Example:

```js
class Foo {
   constructor(a) { console.log(a); }
}
const BoundFoo = Foo.bind(null, 'bound');

new BoundFoo()
```
